データベース設計

FjiiNamiki opened the pull request:
# What
ER図でテーブル間の関係性を図式化
ER図のURL：https://gyazo.com/5ce357837d474b77d1ee9cc013734fb4
READMEの作成

# Why
データベース設計の内容を反映させるため。

FjiiNamiki committed:
データベース設計

yuri240 commented on README.md:
```diff
@@ -22,3 +22,95 @@ Things you may want to cover:
+| firstname_kanji    | string | null: false |
+| lastname_kana      | string | null: false |
+| firstname_kana     | string | null: false |
+| birthday           | string | null: false |
```
誕生日のカラムですが、「date型」を用いることで年、月、日を1つにして保存できます。

stringからdateに変更しましょう！

yuri240 commented on README.md:
```diff
@@ -22,3 +22,95 @@ Things you may want to cover:
+| firstname_kana     | string | null: false |
+| birthday           | string | null: false |
+| email              | string | null: false |
+| password           | string | null: false |
```
```
| password        | string | null: false |
```
こちらのカラムはdeviseのgemを使用しての実装になるため、deviseが予め用意している
`encrypted_password`というカラム名に変更しましょう。

yuri240 commented on README.md:
```diff
@@ -22,3 +22,95 @@ Things you may want to cover:
+
+| Column             | Type            | Options                        |
+| ------------------ | --------------- | ------------------------------ |
+| image              | ActiveStorage   | null: false                    |
```
imageカラムですが、

今回の実装では商品出品の画像投稿において「active_storage」を使用するため、
設計の段階から削除しておきましょう。

理由：active_storage導入時に自動でテーブルなどが生成されるからです。

yuri240 commented on README.md:
```diff
@@ -22,3 +22,95 @@ Things you may want to cover:
+
+| Column             | Type            | Options                        |
+| ------------------ | --------------- | ------------------------------ |
+| zipcode            | integer         | null: false                    |
```
郵便番号はハイフンを必要とすることも考慮して**string型**で実装しましょう！

yuri240 commented on README.md:
```diff
@@ -22,3 +22,95 @@ Things you may want to cover:
+| adress             | integer         | null: false                    |
+| condition_id       | integer         | null: false                    |
+| building           | string          |                                |
+| phonenumber        | integer         | null: false                    |
```
```
| phonenumber        | integer         | null: false                    |
```
電話番号のように先頭が0で始まるものは、
integerとしますと0落ちという先頭の0が消える現象が起きる可能性があります。

したがって、**string型**に変更しましょう！

yuri240 commented on README.md:
```diff
@@ -22,3 +22,95 @@ Things you may want to cover:
+| zipcode            | integer         | null: false                    |
+| prefecture_id      | integer         | null: false                    |
+| city               | string          | null: false                    |
+| adress             | integer         | null: false                    |
```
adressカラムですが、
こちらは「番地」を入力する項目でお間違い無いでしょうか？

番地は、文字列が含まれる場合もございますので
**string型**で設計しましょう！

yuri240 commented on README.md:
```diff
@@ -22,3 +22,95 @@ Things you may want to cover:
+| prefecture_id      | integer         | null: false                    |
+| city               | string          | null: false                    |
+| adress             | integer         | null: false                    |
+| condition_id       | integer         | null: false                    |
```
```
| condition_id       | integer         | null: false                    |
```
配送先住所を入力する際に、
上記カラムは不要と思われます。

したがって、削除していただくようお願いいたします。

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
７点コメントさせていただきましたので、ご確認お願いいたします。

※また、ER図も修正が必要となりますので、お手数ではありますが
修正後のER図を再度こちらのプルリクエストに添付していただくようお願いいたします。

FjiiNamiki committed:
データベース設計_修正1

誕生日のカラムをstringからdateに変更。
zipcodeをintegerからstringに変更。
passwordをencrypted_passwordというカラム名に変更。
imageカラムを削除。
phonenumberをintegerからstringに変更。
adressをintegerからstringに変更。
discriptionsテーブルからcondition_idを削除。

FjiiNamiki commented:
修正しました。
# What
1.誕生日のカラムをstringからdateに変更。
2.zipcodeをintegerからstringに変更。
3.passwordをencrypted_passwordというカラム名に変更。
4.imageカラムを削除。
5.phonenumberをintegerからstringに変更。
6.adressをintegerからstringに変更。
7.discriptionsテーブルからcondition_idを削除。
ER図の修正版：https://gyazo.com/d444c5f562eb0b0c744814e62ad4fdc6

# Why
1.date型を使うことで年月日をまとめて登録できるから。
2.郵便番号はハイフンを入れるから。
3.deviseが予め用意しているものがあるから。
4.Active_storage導入時に自動的にテーブルが生成されるから。
5.先頭が0になるものはintegerだと0が脱落する恐れがあり、それを防ぐから。
6.番地は文字を含むから。
7.必要のないカラムが混在していたから。

Katsuno0814 commented on README.md:
```diff
@@ -34,9 +34,9 @@ Things you may want to cover:
 | firstname_kana     | string | null: false |
-| birthday           | string | null: false |
+| birthday           | date   | null: false |
 | email              | string | null: false |
```
| email | string | null: false |
下記のユーザー管理機能の実装条件に合うように、Optionsに記述も追加しましょう。

- メールアドレスが一意性であること

Katsuno0814 commented on README.md:
```diff
@@ -50,7 +50,6 @@ Things you may want to cover:
-| image              | ActiveStorage   | null: false                    |
 | title              | string          | null: false                    |
 | discription        | text            | null: false                    |
 | category_id        | integer         | null: false, ActiveHash        |
```
>| category_id        | integer         | null: false, ActiveHash        |
| condition_id       | integer         | null: false, ActiveHash        |
| shipping_pay_id    | integer         | null: false, ActiveHash        |
| prefecture_id      | integer         | null: false, ActiveHash        |
| transport_days_id  | integer         | null: false, ActiveHash        |

こちらOptionsに記載がありますActiveHashの記述は不要になりますので削除しておきましょう！

Katsuno0814 replied:
また以下のカラムですが transport_daysと複数形になっているかと思われます。
```
| transport_days_id  | integer         | null: false, ActiveHash        |
```
activehashのカラム名は外部キーカラムと同じく「単数形+_id」になります。
transport_daysで単数形と捉えて実装する場合は問題ありませんが、もし期待通りでない場合はカラム名を transport_day_idにしていただければと思います。

Katsuno0814 commented on README.md:
```diff
@@ -86,13 +85,12 @@ Things you may want to cover:
-| phonenumber        | integer         | null: false                    |
+| phonenumber        | string          | null: false                    |
 | order              | references      | null: false, foreign_key: true |
 | user               | references      | null: false, foreign_key: true |
```
>| user               | references      | null: false, foreign_key: true |

今回の設計では「配送先住所と購入管理が紐づく」実装を行なっていきたいので、ユーザーの外部キーカラムやアソシエーションは不要になります！
修正いたしましょう！

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
４点コメントいたしましたので、ご確認お願いします！

修正頂いた後、コミットプッシュを確認いただき、再度レビュー依頼のご提出をお願い致します！

FjiiNamiki committed:
データベース設計修正2

emailカラムのオプションにunique: trueを追記。
ActiveHashをオプションから削除。
transport_days_idカラムのカラム名をtransport_day_idに変更。
distinationsテーブルとusersテーブルのアソシエーションを削除。

FjiiNamiki commented:
レビューありがとうございます。修正しました。

# What
1.emailカラムのオプションにunique: trueを追記。
2.ActiveHashをオプションから削除。
3.transport_days_idカラムのカラム名をtransport_day_idに変更。
4.distinationsテーブルとusersテーブルのアソシエーションを削除。
ER図の修正版：https://gyazo.com/d8e4d2f551ea3ab5744bdef18e477ea3

# Why
1.メールアドレスの登録に一意性制約をかけるため。
2.オプションへの記載は不要だから。
3.「単数形＋_id」とするのが一般的だから。
4.配送先住所とユーザーにはアソシエーションがないから。

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます。

修正点確認いたしました。
LGTMとさせていただきます。実装お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き実装頑張っていきましょう。
